storage/Utils/TopologyImpl.cc

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (c) 2016 SUSE LLC
+ *
+ * All Rights Reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of version 2 of the GNU General Public License as published
+ * by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, contact Novell, Inc.
+ *
+ * To contact Novell about this file by physical or electronic mail, you may
+ * find current contact information at www.novell.com.
+ */
+
+
+#include <iostream>
+
+#include "storage/Utils/TopologyImpl.h"
+
+
+namespace storage
+{
+
+    void
+    Topology::Impl::set_alignment_offset(long alignment_offset)
+    {
+	Impl::alignment_offset = alignment_offset;
+    }
+
+
+    void
+    Topology::Impl::set_optimal_io_size(unsigned long optimal_io_size)
+    {
+	Impl::optimal_io_size = optimal_io_size;
+    }
+
+
+    unsigned long
+    Topology::Impl::calculate_grain() const
+    {
+	unsigned long grain = max(optimal_io_size, 1UL);
+
+	while (grain < 1024 * 1024)
+	    grain *= 2;
+
+	return grain;
+    }
+
+
+    unsigned long long
+    Topology::Impl::align(unsigned long long sector, unsigned long block_size, Location location) const
+    {
+	long alignment_offset_in_blocks = alignment_offset / block_size;
+	unsigned long grain_in_blocks = calculate_grain() / block_size;
+
+	unsigned long long ret = sector - alignment_offset_in_blocks;
+
+	long rest = ret % grain_in_blocks;
+	if (rest != 0)
+	{
+	    switch (location)
+	    {
+		case Location::START:
+		    ret += grain_in_blocks - rest;
+		    break;
+
+		case Location::END:
+		    ret -= rest + 1;
+		    break;
+	    }
+	}
+
+	ret += alignment_offset_in_blocks;
+
+	return ret;
+    }
+
+
+    // TODO aligned region can be non-existent, return length = 0 or exception?
```

Comment:
IMHO exception. We are using exceptions for all kinds of much minor stuff lin libstorage-ng, and this is truly an exceptional situation.